core/src/main/java/ai/timefold/solver/core/impl/domain/entity/descriptor/EntityDescriptor.java

```diff
@@ -913,6 +901,74 @@ public boolean hasNoNullVariables(Object entity) {
         };
     }
 
+    private Predicate<Object> forEachIncludingPredicateBasicVarSupplier(EnumSet<ForEachInclude> includeSet) {
+        // TODO: Use a supply for this?
+        final var inconsistentDescriptor = this.getShadowVariablesInconsistentDescriptor();
+        final Predicate<Object> isUnassignedPredicate = this::hasNoNullVariables;
+        final Predicate<Object> isInconsistentPredicate = (inconsistentDescriptor == null) ? ignored -> true
+                : entity -> Boolean.FALSE.equals(inconsistentDescriptor.getValue(entity));
+
+        if (includeSet.isEmpty()) {
+            return entity -> isUnassignedPredicate.test(entity) && isInconsistentPredicate.test(entity);
+        }
+
+        var excludeToPredicate = Map.of(
+                ForEachInclude.UNASSIGNED, isUnassignedPredicate,
+                ForEachInclude.INCONSISTENT, isInconsistentPredicate);
+
+        return getForEachPredicateIncluding(includeSet, excludeToPredicate);
+    }
+
+    private Predicate<Object> forEachIncludingPredicateListVarSupplier(EnumSet<ForEachInclude> includeSet) {
+        var listVariableDescriptor = solutionDescriptor.getListVariableDescriptor();
+        if (listVariableDescriptor == null || !listVariableDescriptor.acceptsValueType(entityClass)) {
+            throw new IllegalStateException(
+                    "Impossible state: method called without an applicable list variable descriptor.");
+        }
+        var applicableShadowDescriptor = listVariableDescriptor.getInverseRelationShadowVariableDescriptor();
+        if (applicableShadowDescriptor == null && !includeSet.contains(ForEachInclude.UNASSIGNED)) {
+            throw new IllegalStateException(
+                    "Impossible state: method called without an applicable list variable descriptor.");
+        }
+
+        // This predicate is only used when includeSet does not contain ForEachInclude.UNASSIGNED
+        // so applicableShadowDescriptor cannot be null.
+        final Predicate<Object> isUnassignedPredicate = getForEachFilterPredicateIncludingBasicVar(EnumSet.complementOf(
+                EnumSet.of(ForEachInclude.UNASSIGNED)).toArray(ForEachInclude[]::new))
+                .and(entity -> applicableShadowDescriptor.getValue(entity) != null);
+        final Predicate<Object> isInconsistentPredicate = getForEachFilterPredicateIncludingBasicVar(
+                EnumSet.complementOf(
+                        EnumSet.of(ForEachInclude.INCONSISTENT)).toArray(ForEachInclude[]::new));
```

Comment:
We don't use final local variables, unless lambdas force us to, which I don't think is the case here.